util/chunk/row_container_reader.go

```diff
@@ -0,0 +1,169 @@
+	iter.rc.m.RLock()
+	defer iter.rc.m.RUnlock()
+
+	if !iter.rc.alreadySpilled() {
+		// just any long enough channel should be fine.
+		iter.rowCh = make(chan Row, 512)
+	} else if iter.rc.NumChunks() == 0 {
+		iter.rowCh = make(chan Row)
+	} else {
+		// TODO: give a more reasonable assumption of chunksize
+		assumeChunkSize := iter.rc.m.records.inDisk.numRowsOfEachChunk[0]
+		// To avoid blocking in sending to `rowCh` and  don't start reading the next chunk, it'd be better to give it
+		// a buffer with double size of chunk.
+		iter.rowCh = make(chan Row, 2*assumeChunkSize)
+	}
+}
+
+func (iter *forwardOnlyIterator4RowContainer) dumpChunk(chkIdx int) {
+	iter.rc.m.RLock()
+	defer iter.rc.m.RUnlock()
+
+	if !iter.rc.alreadySpilled() {
+		for rowIdx := 0; rowIdx < iter.rc.NumRowsOfChunk(chkIdx); rowIdx++ {
+			row, err := iter.rc.GetRow(RowPtr{uint32(chkIdx), uint32(rowIdx)})
+			if err != nil {
+				iter.err = err
+				break
+			}
+
+			iter.rc.m.RUnlock()
+			iter.rowCh <- row
+			iter.rc.m.RLock()
+		}
+	} else {
+		l := iter.rc.m.records.inDisk
+
+		// This implementation assumes that all rows in a chunk is stored together tightly.
+		chk := NewChunkWithCapacity(l.fieldTypes, l.NumRowsOfChunk(chkIdx))
+		chkSize := l.numRowsOfEachChunk[chkIdx]
+
+		firstRowOffset, err := l.getOffset(uint32(chkIdx), 0)
+		if err != nil {
+			iter.err = err
+			return
+		}
+
+		formatCh := make(chan rowInDisk, chkSize)
+		go func() {
+			defer close(formatCh)
+
+			iter.rc.m.RLock()
+			defer iter.rc.m.RUnlock()
+
+			// If the row is small, a bufio can significantly improve the performance. As benchmark shows, it's still not bad
+			// for longer rows.
+			r := bufio.NewReader(l.dataFile.getSectionReader(firstRowOffset))
+			format := rowInDisk{numCol: len(l.fieldTypes)}
+			for rowIdx := 0; rowIdx < chkSize; rowIdx++ {
+				_, err = format.ReadFrom(r)
+				if err != nil {
+					// it's safe to set error here. Nothing else will modify the error concurrently.
+					// once this loop breaks, the outer one will also stop after draining the channel
+					iter.err = err
+					break
+				}
+
+				iter.rc.m.RUnlock()
+				formatCh <- format
+				iter.rc.m.RLock()
+			}
+		}()
+
+		rows := make([]Row, 0, chkSize)
+		for format := range formatCh {
+			var row Row
+			row, chk = format.toRow(l.fieldTypes, chk)
+			rows = append(rows, row)
+		}
+
+		for _, row := range rows {
+			iter.rc.m.RUnlock()
+			iter.rowCh <- row
+			iter.rc.m.RLock()
+		}
+	}
+}
+
+func (iter *forwardOnlyIterator4RowContainer) startWorker() {
+	for chkIdx := 0; chkIdx < iter.rc.NumChunks(); chkIdx++ {
+		iter.dumpChunk(chkIdx)
+	}
+
+	close(iter.rowCh)
+}
+
+// NewForwardOnlyIterator4RowContainer creates a forward only iterator for row container
+func NewForwardOnlyIterator4RowContainer(rc *RowContainer) *forwardOnlyIterator4RowContainer {
+	iter := &forwardOnlyIterator4RowContainer{
+		rc: rc,
+	}
+	iter.initializeChannel()
+	go iter.startWorker()
```

Comment:
This goroutine leaks. I'll need to add a method to close it. (Or simply observe the status of rowContainer and close it automatically, to keep the interface simpler).